sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUuidPathColumnOnProjects.java

```diff
@@ -0,0 +1,156 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.db.version.v60;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.db.Database;
+import org.sonar.db.version.BaseDataChange;
+import org.sonar.db.version.MassUpdate;
+import org.sonar.db.version.Select;
+import org.sonar.db.version.SqlStatement;
+
+import static java.util.stream.Collectors.toCollection;
+
+public class PopulateUuidPathColumnOnProjects extends BaseDataChange {
+
+  private static final Logger LOG = Loggers.get(PopulateUuidPathColumnOnProjects.class);
+  private static final Joiner PATH_JOINER = Joiner.on('.');
+  private static final Splitter PATH_SPLITTER = Splitter.on('.').omitEmptyStrings();
+  private static final String PATH_SEPARATOR = ".";
+  private static final String ROOT_PATH = PATH_SEPARATOR;
+
+  public PopulateUuidPathColumnOnProjects(Database db) {
+    super(db);
+  }
+
+  @Override
+  public void execute(Context context) throws SQLException {
+    // group upgrades by tree of component
+    List<String> rootComponentUuids = context
+      .prepareSelect("select distinct project_uuid from projects where uuid_path is null")
+      .list(row -> row.getString(1));
+    for (String rootUuid : rootComponentUuids) {
+      handleRoot(rootUuid, context);
+    }
+
+    handleOrphans(context);
+  }
+
+  private void handleRoot(String rootComponentUuid, Context context) throws SQLException {
+    Relations relations = new Relations();
+    context
+      .prepareSelect("select s.id, s.path, s.component_uuid from snapshots s where s.root_component_uuid=? and s.islast=?")
+      .setString(1, rootComponentUuid)
+      .setBoolean(2, true)
+      .scroll(row -> {
+        long snapshotId = row.getLong(1);
+        String snapshotPath = row.getString(2);
+        String componentUuid = row.getString(3);
+        relations.add(new Snapshot(snapshotId, snapshotPath, componentUuid));
+      });
+
+    MassUpdate massUpdate = context.prepareMassUpdate();
+    massUpdate.select("select p.uuid, p.project_uuid from projects p where p.project_uuid=? and p.uuid_path is null").setString(1, rootComponentUuid);
+    massUpdate.update("update projects set uuid_path=? where uuid=? and uuid_path is null");
+    massUpdate.rowPluralName("components");
```

Comment:
what about `"uuid_path of components in tree of " + rootComponentUuid` ?
would avoid multiple logs with same label during this migration